src/qibo/gates/gates.py

```diff
@@ -2328,14 +2328,31 @@ def __init__(
             "trainable": trainable,
         }
 
-        # checking unitarity without invoking any backend
         if check_unitary:
-            product = np.transpose(np.conj(unitary)) @ unitary
-            sums = all(np.abs(1 - np.sum(product, axis=1)) < PRECISION_TOL)
-            diagonal = all(np.abs(1 - np.diag(product)) < PRECISION_TOL)
+            if unitary.__class__.__name__ == "Tensor":
+                import torch  # pylint: disable=C0145
+
+                diag_function = torch.diag
+                all_function = torch.all
+                conj_function = torch.conj
+                transpose_function = torch.transpose
+            else:
+                diag_function = np.diag
+                all_function = np.all
+                conj_function = np.conj
+                transpose_function = np.transpose
+
+            product = transpose_function(conj_function(unitary), (1, 0)) @ unitary
+            diagonals = all(np.abs(1 - diag_function(product)) < PRECISION_TOL)
+            off_diagonals = bool(
+                all_function(
+                    np.abs(product - diag_function(diag_function(product)))
+                    < PRECISION_TOL
+                )
+            )
```

Comment:
I believe that here the only function causing problems is `all`, but for both `torch.tensor` and `np.array` you can call `x.all()` rather than explicitely `np/torch.all(x)`. 
Therefore if you do:
```
off_diagonals = bool(
                    (np.abs(product - np.diag(np.diag(product)))
                    < PRECISION_TOL).all()
                )
```
it should work for both

Reply:
I'm not sure `tensorflow.EagerTensor` has an `all()` method